src/rulesets/Base/array.jl

```diff
@@ -1,3 +1,32 @@
+#####
+##### constructors
+#####
+
+ChainRules.@non_differentiable (::Type{T} where {T<:Array})(::UndefInitializer, args...)
+
+function rrule(::Type{T}, x::AbstractArray) where {T<:Array}
+    project_x = ProjectTo(x)
+    Array_pullback(ȳ) = (NoTangent(), project_x(ȳ))
+    return T(x), Array_pullback
+end
+
+#####
+##### `vect`
+#####
+
+@non_differentiable Base.vect()
+
+# Don't worry about projection here. The data passes straight through, so if a cotangent has
+# the wrong type for some reason, it must be the fault of another rule somewhere.
+function rrule(::typeof(Base.vect), X::T...) where {T}
```

Comment:
Why the `where T` here? Is it about avoiding promotion? Of so perhaps worth adding a comment?

Reply:
I guess not because, T could be an abstract type and we test on non-homogenous inputs

Isn't #492 still an issue here?

Reply:
I thought it was there to enforce that this is the case of uniform types only. 
```
julia> ff(xs::T...) where {T} = "ok";

julia> ff(1, 3.4)
ERROR: MethodError: no method matching ff(::Int64, ::Float64)
```

Reply:
That's what I thought too, but then I'm confused about the "inhomogeneous" test

Reply:
Oh that test fails with a MethodError, so that resolves my confusion 😆

Reply:
Good point, but I don't think it does. This is, I believe, equivalent do doing diagonal dispatch, so everything has to be of the same concrete type for this method to be dispatched to (I had to double check though).

For example
```julia
function foo(X::T...) where {T}
    return X
end
foo(5.0, 4)
```
yields
```julia
ERROR: MethodError: no method matching foo(::Float64, ::Int64)
Closest candidates are:
  foo(::T...) where T at REPL[10]:1
Stacktrace:
 [1] top-level scope
   @ REPL[13]:1
```

Reply:
I should have refreshed the page before posting 😂